src/storage/src/hummock/block_stream.rs

```diff
@@ -0,0 +1,239 @@
+    ) -> Self {
+        Self {
+            buf_reader: byte_stream,
+            block_idx: 0,
+            block_metas,
+            buf: Bytes::default(),
+            buff_offset: 0,
+        }
+    }
+
+    /// Reads the next block from the stream and returns it. Returns `None` if there are no blocks
+    /// left to read.
+    pub async fn next_block_impl(&mut self) -> HummockResult<Option<(Bytes, usize)>> {
+        if self.block_idx >= self.block_metas.len() {
+            return Ok(None);
+        }
+
+        let block_meta = &self.block_metas[self.block_idx];
+        fail_point!("stream_read_err", |_| Err(ObjectError::internal(
+            "stream read error"
+        )
+        .into()));
+        let uncompressed_size = block_meta.uncompressed_size as usize;
+        let end = self.buff_offset + block_meta.len as usize;
+        let data = if end > self.buf.len() {
+            let current_block = self.read_next_buf(block_meta.len as usize).await?;
+            self.buff_offset = 0;
+            current_block
+        } else {
+            let data = self.buf.slice(self.buff_offset..end);
+            self.buff_offset = end;
+            data
+        };
+
+        self.block_idx += 1;
+        Ok(Some((data, uncompressed_size)))
+    }
+
+    async fn read_next_buf(&mut self, read_size: usize) -> HummockResult<Bytes> {
+        let mut read_buf = BytesMut::with_capacity(read_size);
+        let start_pos = if self.buff_offset < self.buf.len() {
+            read_buf.extend_from_slice(&self.buf[self.buff_offset..]);
+            self.buf.len() - self.buff_offset
+        } else {
+            0
+        };
+        let mut rest = read_size - start_pos;
+        while rest > 0 {
+            let next_packet = self
+                .buf_reader
+                .read_bytes()
+                .await
+                .unwrap_or_else(|| Err(ObjectError::internal("read unexpected EOF")))?;
+            let read_len = std::cmp::min(next_packet.len(), rest);
+            read_buf.extend_from_slice(&next_packet[..read_len]);
+            rest -= read_len;
+            if rest == 0 {
+                self.buf = next_packet.slice(read_len..);
+                return Ok(read_buf.freeze());
+            }
+        }
+        self.buf = Bytes::default();
+        Ok(read_buf.freeze())
+    }
+
+    pub fn next_block_index(&self) -> usize {
+        self.block_idx
+    }
+
+    pub async fn next_block(&mut self) -> HummockResult<Option<Box<Block>>> {
+        match self.next_block_impl().await? {
+            None => Ok(None),
+            Some((buf, uncompressed_size)) => {
+                Ok(Some(Box::new(Block::decode(buf, uncompressed_size)?)))
+            }
+        }
+    }
+}
+
+/// An iterator that reads the blocks of an SST step by step from a given stream of bytes optimize for batch query.
+/// Because all data in batch query shall be put into block-cache, we can not assign a reference of Bytes for each block.
```

Comment:
Because if we do not copy this block and put it to block-cache, it may hold the whole large memory.
So if we do not fill it back to cache, it is ok to use Bytes clone.